File: benches/ports/js/accum.js
// accum rows at n = 1 000 000, ported from acvus-interpreter-test/benches/accum.rs.
//
// Obligation across artifacts: each row here is the semantics of that file's
// `rust_*` twin, which is what the bench's `rust/us` column times. Every value
// the rows carry is below 2^53, so a JS double holds each of them exactly;
// `black_box` has no JS spelling, so where V8 hoists work the twin is stopped
// from hoisting, the row says so in docs/performance.md.
//
// Usage: node accum.js <row> [n] [reps]

const REPS = 4;
const N = 1000000;

function idOf(i) {
  return i;
}

function evenOf(i) {
  return i % 2 === 0;
}

function step(x) {
  return x + 1;
}

function* range(n) {
  for (let i = 0; i < n; i++) yield i;
}

function intWhile(n) {
  let acc = 0;
  let i = 0;
  while (i < n) {
    acc = acc + i;
    i = i + 1;
  }
  return acc;
}

function floatWhile(n) {
  let acc = 0.0;
  let i = 0;
  while (i < n) {
    acc = acc + i;
    i = i + 1;
  }
  return acc;
}

function rangeSum(n) {
  return range(n).reduce((a, b) => a + b, 0);
}

function mapIdSum(n) {
  return range(n)
    .map((x) => x)
    .reduce((a, b) => a + b, 0);
}

function mapAddSum(n) {
  return range(n)
    .map((x) => x + 1)
    .reduce((a, b) => a + b, 0);
}

function branchWhile(n) {
  let acc = 0;
  let i = 0;
  while (i < n) {
    if (evenOf(i)) acc = acc + i;
    i = i + 1;
  }
  return acc;
}

function callWhile(n) {
  let i = 0;
  while (i < n) i = step(i);
  return i;
}

function collatzWhile(n) {
  let acc = 0;
  let i = 0;
  while (i < n) {
    const d = i % 2 === 0 ? i / 2 : i * 3 + 1;
    acc = acc + d;
    i = i + 1;
  }
  return acc;
}

function gradeWhile(n) {
  let a = 0;
  let b = 0;
  let i = 0;
  while (i < n) {
    if (i % 3 === 0) a = a + 1;
    else if (i % 3 === 1) b = b + 1;
    else a = a + 2;
    i = i + 1;
  }
  return a + b;
}

function whileLetVec(n) {
  const v = [];
  for (let i = 0; i < n; i++) v.push(i);
  const it = v[Symbol.iterator]();
  let acc = 0;
  let next = it.next();
  while (!next.done) {
    acc = acc + next.value;
    next = it.next();
  }
  return acc;
}

function forRange(n) {
  let acc = 0;
  for (let i = 0; i < n; i++) acc = acc + i;
  return acc;
}

function forSliceAdd(n) {
  const v = [];
  for (let i = 0; i < n; i++) v.push(i);
  let acc = 0;
  for (const x of v) acc = acc + x + 1;
  return acc;
}

const ROWS = {
  "int while": intWhile,
  "float while": floatWhile,
  "range | sum": rangeSum,
  "map id | sum": mapIdSum,
  "map add | sum": mapAddSum,
  "branch while": branchWhile,
  "call while": callWhile,
  "collatz while": collatzWhile,
  "grade while": gradeWhile,
  "while let vec": whileLetVec,
  "for range": forRange,
  "for slice add": forSliceAdd,
};

const row = process.argv[2];
const n = process.argv[3] === undefined ? N : Number(process.argv[3]);
const reps = process.argv[4] === undefined ? REPS : Number(process.argv[4]);
const body = ROWS[row];
if (body === undefined) {
  console.error(`no row named ${JSON.stringify(row)}`);
  process.exit(1);
}

body(n);
const samples = [];
let value;
for (let r = 0; r < reps; r++) {
  const start = process.hrtime.bigint();
  value = body(n);
  samples.push(Number(process.hrtime.bigint() - start));
}
samples.sort((a, b) => a - b);
const median = samples[samples.length >> 1];
const shown = row === "float while" ? value.toFixed(1) : String(value);
console.log(`${row}\t${(median / 1000).toFixed(1)}\t${shown}`);

File: benches/ports/python/accum.py
"""accum rows at n = 1 000 000, ported from acvus-interpreter-test/benches/accum.rs.

Obligation across artifacts: each row here is the semantics of that file's
`rust_*` twin, which is what the bench's `rust/us` column times and what its
assertion holds the acvus script to. `black_box` has no Python spelling, so
what the twin is stopped from folding, CPython cannot fold anyway.

Usage: accum.py <row> [n] [reps]
"""

import sys
import time

REPS = 4
N = 1_000_000


def id_of(i):
    return i


def even_of(i):
    return i % 2 == 0


def step(x):
    return x + 1


def int_while(n):
    acc = 0
    i = 0
    while i < n:
        acc = acc + i
        i = i + 1
    return acc


def float_while(n):
    acc = 0.0
    i = 0
    while i < n:
        acc = acc + float(i)
        i = i + 1
    return acc


def range_sum(n):
    return sum(range(n))


def map_id_sum(n):
    return sum(map(lambda x: x, range(n)))


def map_add_sum(n):
    return sum(map(lambda x: x + 1, range(n)))


def branch_while(n):
    acc = 0
    i = 0
    while i < n:
        if even_of(i):
            acc = acc + i
        i = i + 1
    return acc


def call_while(n):
    i = 0
    while i < n:
        i = step(i)
    return i


def collatz_while(n):
    acc = 0
    i = 0
    while i < n:
        d = i // 2 if i % 2 == 0 else i * 3 + 1
        acc = acc + d
        i = i + 1
    return acc


def grade_while(n):
    a = 0
    b = 0
    i = 0
    while i < n:
        if i % 3 == 0:
            a = a + 1
        elif i % 3 == 1:
            b = b + 1
        else:
            a = a + 2
        i = i + 1
    return a + b


_EXHAUSTED = object()


def while_let_vec(n):
    v = list(range(n))
    it = iter(v)
    acc = 0
    while True:
        x = next(it, _EXHAUSTED)
        if x is _EXHAUSTED:
            break
        acc = acc + x
    return acc


def for_range(n):
    acc = 0
    for i in range(n):
        acc = acc + i
    return acc


def for_slice_add(n):
    v = list(range(n))
    acc = 0
    for x in v:
        acc = acc + x + 1
    return acc


ROWS = {
    "int while": int_while,
    "float while": float_while,
    "range | sum": range_sum,
    "map id | sum": map_id_sum,
    "map add | sum": map_add_sum,
    "branch while": branch_while,
    "call while": call_while,
    "collatz while": collatz_while,
    "grade while": grade_while,
    "while let vec": while_let_vec,
    "for range": for_range,
    "for slice add": for_slice_add,
}


def main():
    row = sys.argv[1]
    n = int(sys.argv[2]) if len(sys.argv) > 2 else N
    reps = int(sys.argv[3]) if len(sys.argv) > 3 else REPS
    body = ROWS[row]

    body(n)
    samples = []
    value = None
    for _ in range(reps):
        start = time.perf_counter_ns()
        value = body(n)
        samples.append(time.perf_counter_ns() - start)
    samples.sort()
    median = samples[len(samples) // 2]
    shown = repr(value) if isinstance(value, float) else str(value)
    print(f"{row}\t{median / 1000.0:.1f}\t{shown}")


main()
